src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,309 @@
+#include "rgw_url.h"
+#include "rgw_aws.h"
+#include <string>
+#include <iostream>
+#include <utility>
+#include "common/dout.h"
+#include <aws/core/Aws.h>
+#include <aws/sns/SNSClient.h>
+#include <aws/lambda/LambdaClient.h>
+#include <aws/lambda/model/InvokeRequest.h>
+#include <boost/lockfree/queue.hpp>
+#include <aws/core/utils/ARN.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/sns/model/PublishRequest.h>
+
+#define dout_subsys ceph_subsys_rgw
+
+namespace rgw::aws {
+  static const int STATUS_OK = 1;
+  static const int STATUS_FULL = 2;
+  static const int STATUS_MANAGER_STOPPED = 3;
+  static const int STATUS_CLIENT_CLOSED = 4;
+
+  struct client_id_t {
+    const std::string accessKey;
+    const std::string accessSecret;
+    const std::string arn;
+    const bool useSSL;
+
+    client_id_t(const std::string &accessKey, const std::string &accessSecret,
+                const std::string &arn, bool useSsl) : accessKey(accessKey),
+                                                      accessSecret(accessSecret),
+                                                      arn(arn), useSSL(useSsl) {}
+
+    bool operator==(const client_id_t &other) const {
+      return accessKey == other.accessKey &&
+             accessSecret == other.accessSecret &&
+             arn == other.arn && useSSL == other.useSSL;
+    }
+
+    struct hasher {
+      std::size_t operator()(const client_id_t &k) const {
+        return ((std::hash<std::string>()(k.accessKey)
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+  class AsyncCallerContext: public Aws::Client::AsyncCallerContext{
+  public:
+    const reply_callback_t cb;
+    CephContext* const cct;
+    AsyncCallerContext(reply_callback_t cb, CephContext* cct):cb(cb), cct(cct){}
+  };
+
+  void invoke_callback(const Aws::Lambda::LambdaClient* client,
+                              const Aws::Lambda::Model::InvokeRequest& request,
+                              const Aws::Lambda::Model::InvokeOutcome & outcome,
+                              const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context){
+
+    auto ctx = (AsyncCallerContext*) context.get();
+    CephContext* cct = ctx->cct;
+    if (outcome.IsSuccess()) {
+      ldout(cct, 20) << "Invoke function: " << request.GetFunctionName() << " Response: " << outcome.GetResult().GetLogResult() << dendl;
+      if(ctx && ctx->cb) {
+        ctx->cb(STATUS_OK);
+      }
+    }
+    else {
+      const auto& error = outcome.GetError();
+      ldout(cct, 20) << "ERROR: " << error.GetExceptionName() << ": "
+                << error.GetMessage() << dendl;
+      if(ctx && ctx->cb) {
+        ctx->cb(-(int)error.GetErrorType());
```

Comment:
do you get an error if you dont cast to int?
don't think this is needed when casting from enum.

Reply:
https://stackoverflow.com/questions/11421432/how-can-i-output-the-value-of-an-enum-class-in-c11
I had to use static_cast.

Reply:
ok. so please use static_cast instead of c-style case